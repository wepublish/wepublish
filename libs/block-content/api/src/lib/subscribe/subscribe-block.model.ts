import {Field, InputType, ObjectType, OmitType} from '@nestjs/graphql'
import {BaseBlock} from '../base-block.model'
import {BlockType} from '../block-type.model'

@ObjectType({
  implements: BaseBlock,
})
export class SubscribeBlock extends BaseBlock<typeof BlockType.Subscribe> {
  @Field(() => [String], {nullable: true})
  memberPlanIds?: string[]
}

@InputType()
export class SubscribeBlockInput extends OmitType(
  SubscribeBlock,
  ['type'] as const,
  InputType
) {}
