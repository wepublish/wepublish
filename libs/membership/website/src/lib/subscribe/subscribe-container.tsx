import {MutationResult, QueryResult} from '@apollo/client'
import {useUser} from '@wepublish/authentication/website'
import {useWebsiteBuilder} from '@wepublish/website/builder'
import {
  ChallengeQuery,
  MemberPlanListQuery,
  SubscribeMutation,
  useChallengeQuery,
  useMemberPlanListQuery,
  useSubscribeMutation
} from '@wepublish/website/api'
import {useEffect} from 'react'

export type SubscribeContainerProps = {
  onSubscribeMutation?: (
    mutationResult: Pick<MutationResult<SubscribeMutation>, 'data' | 'loading' | 'error'>
  ) => void

  onChallengeQuery?: (
    queryResult: Pick<QueryResult<ChallengeQuery>, 'data' | 'loading' | 'error' | 'refetch'>
  ) => void

  onMemberPlansQuery?: (
    queryResult: Pick<QueryResult<MemberPlanListQuery>, 'data' | 'loading' | 'error' | 'refetch'>
  ) => void
}

export const SubscribeContainer = ({
  onChallengeQuery,
  onMemberPlansQuery,
  onSubscribeMutation
}: SubscribeContainerProps) => {
  const {hasUser, setToken} = useUser()
  const {Subscribe} = useWebsiteBuilder()

  const challenge = useChallengeQuery()
  const memberPlanList = useMemberPlanListQuery({
    variables: {
      take: 50
    }
  })

  const [subscribe, result] = useSubscribeMutation({
    onError: () => challenge.refetch(),
    onCompleted(data) {
      if (data.registerMemberAndReceivePayment.session) {
        setToken(data.registerMemberAndReceivePayment.session)
      }

      if (data.registerMemberAndReceivePayment.payment.intentSecret) {
        window.location.href = data.registerMemberAndReceivePayment.payment.intentSecret
      }
    }
  })

  useEffect(() => {
    onChallengeQuery?.(challenge)
  }, [challenge, onChallengeQuery])

  useEffect(() => {
    onMemberPlansQuery?.(memberPlanList)
  }, [memberPlanList, onMemberPlansQuery])

  useEffect(() => {
    onSubscribeMutation?.(result)
  }, [result, onSubscribeMutation])

  if (hasUser) {
    return null
  }

  return (
    <Subscribe
      challenge={challenge}
      memberPlans={memberPlanList}
      subscribe={result}
      onSubmit={formData =>
        subscribe({
          variables: {
            ...formData
          }
        })
      }
    />
  )
}
