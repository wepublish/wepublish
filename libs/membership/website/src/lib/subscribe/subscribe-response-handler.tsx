import {FetchResult} from '@apollo/client'
import {StripeElement, StripePayment} from '@wepublish/payment/website'
import {FullMemberPlanFragment, SubscribeMutation, usePageLazyQuery} from '@wepublish/website/api'
import {useWebsiteBuilder} from '@wepublish/website/builder'
import {useCallback, useEffect, useState} from 'react'

interface SubscribeResponseHandlerProps {
  subscribeResponse: FetchResult<SubscribeMutation> | undefined
  memberPlan: FullMemberPlanFragment | undefined
  onError?: () => void
}

/**
 * You can use this component after you called a subscribe mutation.
 * It will handle the necessary payment form, re-directs and error-handling.
 * For an implementation example see subscribe-container.tsx
 * You may want to re-load data via onError prop callback function
 */
export function SubscribeResponseHandler({
  subscribeResponse,
  memberPlan,
  onError
}: SubscribeResponseHandlerProps) {
  const {
    elements: {Alert}
  } = useWebsiteBuilder()

  const [fetchPage] = usePageLazyQuery()

  const [stripeClientSecret, setStripeClientSecret] = useState<string>()
  const [errors, setErrors] = useState<string[] | undefined>()

  const handleError = useCallback(
    (errors: string[]) => {
      setErrors(errors)
      onError?.()
    },
    [onError]
  )

  const handleSubscribeResponse = useCallback(async () => {
    if (!subscribeResponse) {
      return
    }

    const successPage = await fetchPage({
      variables: {
        id: memberPlan?.successPageId
      }
    })
    const successUrl = successPage.data?.page.url || '/profile'

    if (subscribeResponse.errors) {
      handleError(subscribeResponse.errors.map(error => error.message))
      return
    }

    if (!subscribeResponse.data) {
      handleError(['Unerwarteter Fehler: Die Serverantwort war ohne Daten.'])
      return
    }

    const {createSubscription} = subscribeResponse.data

    // subscription was paid with existing credit card.
    // in this case intentSecret can be undefined. So this check must run before checking if intentSecret is empty!
    if (createSubscription.state === 'paid') {
      window.location.href = successUrl
      return
    }

    if (!createSubscription.intentSecret) {
      handleError(['Unerwarteter Fehler: Die Serverantwort lieferte keinen intent secret zurÃ¼ck.'])
      return
    }

    // payrexx
    if (createSubscription.intentSecret.startsWith('http')) {
      window.location.href = createSubscription.intentSecret as string
      return
    } else if (createSubscription.intentSecret.startsWith('no_charge')) {
      // trial subscriptions: non-charge-payment-adapter
      window.location.href = successUrl
      return
    } else if (createSubscription.paymentMethod.paymentProviderID === 'stripe') {
      setStripeClientSecret(createSubscription.intentSecret)
      return
    }

    // handle fallback
    handleError(['Unbekannter Fehler!'])
  }, [subscribeResponse, handleError, memberPlan, fetchPage])

  // entry point. whenever a subscribe response is passed, this component handles it.
  useEffect(() => {
    handleSubscribeResponse()
  }, [subscribeResponse, handleSubscribeResponse])

  async function onCloseStripeElement(success: boolean) {
    if (!memberPlan) {
      window.location.href = '/profile'
      return
    }

    const page = await fetchPage({
      variables: {
        id: success ? memberPlan.successPageId : memberPlan.failPageId
      }
    })

    window.location.href = page.data?.page.url ?? '/profile'
  }

  return (
    <>
      {errors && errors.map(error => <Alert security="warning">{error}</Alert>)}
      {stripeClientSecret && (
        <StripeElement clientSecret={stripeClientSecret}>
          <StripePayment
            onClose={async success => {
              await onCloseStripeElement(success)
            }}
          />
        </StripeElement>
      )}
    </>
  )
}
