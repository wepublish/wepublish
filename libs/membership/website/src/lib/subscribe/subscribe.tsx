import styled from '@emotion/styled'
import {BuilderSubscribeProps} from '@wepublish/website/builder'

export type SubscribeProps = BuilderSubscribeProps

const StyledSubscribe = styled.div`
  color: pink;
`

export function Subscribe({className, challenge}: SubscribeProps) {
  return (
    <StyledSubscribe className={className}>
      <h1>Welcome to Subscribe!</h1>
    </StyledSubscribe>
  )
}
