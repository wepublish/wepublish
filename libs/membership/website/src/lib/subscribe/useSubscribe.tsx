import {usePageLazyQuery} from '@wepublish/website/api'

export function useSubscribe() {
  // @TODO: Replace with objects on Memberplan when Memberplan has been migrated to V2
  // Pages are currently in V2 and Memberplan are in V1, so we have no access to page objects.
  const [fetchPage] = usePageLazyQuery()

  async function fetchRedirectPages({
    successPageId,
    failPageId
  }: {
    successPageId?: string | undefined | null
    failPageId?: string | undefined | null
  }) {
    return await Promise.all([
      fetchPage({
        variables: {
          id: successPageId
        }
      }),
      fetchPage({
        variables: {
          id: failPageId
        }
      })
    ])
  }

  return {
    fetchRedirectPages
  }
}
