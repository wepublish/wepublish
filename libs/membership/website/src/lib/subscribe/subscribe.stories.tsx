import {ComponentStory, Meta} from '@storybook/react'
import {Subscribe} from './subscribe'

export default {
  component: Subscribe,
  title: 'Subscribe'
} as Meta

const Template: ComponentStory<typeof Subscribe> = args => <Subscribe {...args} />
export const Default = Template.bind({})
