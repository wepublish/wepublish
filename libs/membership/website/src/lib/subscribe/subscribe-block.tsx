import {BuilderSubscribeBlockProps} from '@wepublish/website/builder'

import {SubscribeContainer} from './subscribe-container'

export type SubscribeBlockProps = BuilderSubscribeBlockProps

export const SubscribeBlock = ({className, memberPlanIds}: SubscribeBlockProps) => {
  return <SubscribeContainer className={className} memberPlanIds={memberPlanIds ?? undefined} />
}
