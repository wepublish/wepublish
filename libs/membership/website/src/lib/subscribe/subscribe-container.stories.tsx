import {action} from '@storybook/addon-actions'
import {ComponentStory, Meta} from '@storybook/react'
import {ChallengeDocument} from '@wepublish/website/api'
import {SubscribeContainer} from './subscribe-container'

export default {
  component: SubscribeContainer,
  title: 'Container/Subscribe'
} as Meta

const Template: ComponentStory<typeof SubscribeContainer> = args => <SubscribeContainer {...args} />
export const Default = Template.bind({})

Default.args = {
  onChallengeQuery: action('onChallengeQuery'),
  onSubscribeMutation: action('onSubscribeMutation'),
  onMemberPlansQuery: action('onMemberPlansQuery')
}

Default.parameters = {
  apolloClient: {
    mocks: [
      {
        request: {
          query: ChallengeDocument
        },
        result: {
          data: {}
        }
      }
    ]
  }
}

export const WithClassName = Template.bind({})

WithClassName.args = {
  onChallengeQuery: action('onChallengeQuery'),
  onSubscribeMutation: action('onSubscribeMutation'),
  onMemberPlansQuery: action('onMemberPlansQuery'),
  className: 'extra-classname'
}

WithClassName.parameters = {
  apolloClient: {
    mocks: [
      {
        request: {
          query: ChallengeDocument
        },
        result: {
          data: {}
        }
      }
    ]
  }
}
