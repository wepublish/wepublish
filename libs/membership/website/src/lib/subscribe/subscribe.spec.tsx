import {render} from '@testing-library/react'

import {Subscribe} from './subscribe'

describe('Subscribe', () => {
  it('should render successfully', () => {
    const {baseElement} = render(
      <Subscribe challenge={{data: undefined, error: undefined, loading: true}} />
    )
    expect(baseElement).toBeTruthy()
  })
})
