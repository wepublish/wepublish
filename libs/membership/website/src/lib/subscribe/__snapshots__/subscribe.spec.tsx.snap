// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscribe should render Default 1`] = `
<DocumentFragment>
  <div
    class="css-bdz642"
  >
    <h1>
      Welcome to Subscribe!
    </h1>
  </div>
</DocumentFragment>
`;
